operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,418 @@
+func (c *cebManagerFcfs) getCebCopyFromCache(cebName string) (*cilium_v2.CiliumEndpointBatch, error) {
+	if ceb, exists := c.desiredCebs[cebName]; exists {
+		outCeb := new(cilium_v2.CiliumEndpointBatch)
+		ceb.backendMutex.RLock()
+		ceb.ceb.DeepCopyInto(outCeb)
+		ceb.backendMutex.RUnlock()
+		return outCeb, nil
+
+	}
+	return nil, fmt.Errorf("Failed to get CEB from local cache for the cebName: %s", cebName)
+}
+
+// findCeb returns the available cebTracker object and is chosen based on FCFS.
+// if all CEB's reached max capacity or marked for delete. Allocate a new CEB.
+func (c *cebManagerFcfs) findCeb() (ceb *cebTracker) {
+
+	// get first available CEB
+	for _, ceb = range c.desiredCebs {
+		if len(ceb.ceb.Endpoints) >= c.maxCepsInCeb || len(ceb.ceb.Endpoints) == 0 {
+			continue
+		}
+		return
+	}
+
+	// allocate a new cebTracker and return
+	ceb = c.createCeb("")
+	return
+}
+
+// InsertCepInCache is used to Insert CEP in local cache, this may result in Creating a New
+// CEB object or Updating an existing CEB object.
+func (c *cebManagerFcfs) InsertCepInCache(cep *cilium_v2.CoreCiliumEndpoint) string {
+
+	// check the given cep is already exists in any of the CEB.
+	// if yes, Update a ceb with the given cep object.
+	if cebName, exists := c.cacheCepMap.get(GetCepNameFromCCEP(cep)); exists {
+		// add a cep into the ceb
+		addCEPtoCEB(cep, c.desiredCebs[cebName])
+		c.queue.Add(cebName)
+		return cebName
+	}
+
+	// If given cep object isn't packed in any of the CEB. find a new ceb
+	// to pack this cep.
+	cb := c.findCeb()
+
+	// Cache CEP name with newly allocated CEB.
+	c.cacheCepMap.insert(GetCepNameFromCCEP(cep), cb.ceb.GetName())
+
+	// Queue the CEP in CEB
+	addCEPtoCEB(cep, cb)
+	c.queue.Add(cb.ceb.GetName())
+	return cb.ceb.GetName()
+}
+
+// RemoveCepFromCache is used to remove the CEP from local cache, this may result in
+// Updating an existing CEB object.
+func (c *cebManagerFcfs) RemoveCepFromCache(cepName string) {
+
+	log.WithFields(logrus.Fields{
+		"cep-name": cepName,
+	}).Debug("Remove CEP from local cache")
+
+	// Check in local cache, if a given cep is already batched in one of the ceb.
+	// and if exists, delete cep from ceb.
+	if cebName, exists := c.cacheCepMap.get(cepName); exists {
+		var ceb *cebTracker
+		if ceb, exists = c.desiredCebs[cebName]; !exists {
+			log.Info("Valid cep-ceb mapping, but CEB doesn't exist in ceb cache.")
+			return
+		}
+
+		ceb.backendMutex.Lock()
+		defer ceb.backendMutex.Unlock()
+		for i, ep := range ceb.ceb.Endpoints {
+			if GetCepNameFromCCEP(&ep) == cepName {
+				// Insert deleted CoreCEP in removedCeps
+				ceb.removedCeps[GetCepNameFromCCEP(&ep)] = struct{}{}
+				ceb.ceb.Endpoints =
+					append(ceb.ceb.Endpoints[:i],
+						ceb.ceb.Endpoints[i+1:]...)
+				break
+			}
+		}
+		log.WithFields(logrus.Fields{
+			"ceb-name":  cebName,
+			"cep-name":  cepName,
+			"cep-count": len(ceb.ceb.Endpoints),
+		}).Debug("Removed cep from ceb")
+
+		c.queue.Add(cebName)
+	}
+
```

Comment:
Should throw a warning if `!exists`? It is not critical but if this ever happens, something is not right.

Reply:
Hmm, I think it's okay. Wouldn't we end up in having dead code ?

Reply:
It's not dead code :) otherwise you don't need to check for exists in the first place.

Reply:
Count me in for more logging and error checking.